discovery-handler-modules/coap-discovery-handler/src/discovery_utils.rs

```diff
@@ -0,0 +1,122 @@
+use coap::CoAPClient as Client;
+use coap_lite::{CoapRequest, CoapResponse};
+use log::{debug, info};
+#[cfg(test)]
+use mockall::{automock, predicate::*};
+use std::net::{SocketAddr, ToSocketAddrs};
+use std::time::Duration;
+
+type Result<T> = std::io::Result<T>;
+
+#[cfg_attr(test, automock)]
+pub trait CoAPClient: Sized {
+    fn get_with_timeout(&self, url: &str, timeout: Duration) -> Result<CoapResponse>;
+    fn send_all_coap(&self, request: &CoapRequest<SocketAddr>, segment: u8) -> Result<()>;
+    fn set_receive_timeout(&self, dur: Option<Duration>) -> Result<()>;
+    fn receive_from(&self) -> Result<(CoapResponse, SocketAddr)>;
+}
+
+pub struct CoAPClientImpl {
+    client: Client,
+}
+
+impl CoAPClientImpl {
+    pub fn new<A: ToSocketAddrs>(addr: A) -> Self {
+        CoAPClientImpl {
+            client: Client::new(addr).unwrap(),
+        }
+    }
+}
+
+impl CoAPClient for CoAPClientImpl {
+    fn get_with_timeout(&self, url: &str, timeout: Duration) -> Result<CoapResponse> {
+        Client::get_with_timeout(url, timeout)
```

Comment:
You're right, I should wrap it into a tokio task and use a separate thread. On the other hand, the `async-coap` crate is basically dead, unfortunately. Maybe one day I'll write an async version myself. Most of the protocol implementation has been already extracted to [coap-lite](https://github.com/martindisch/coap-lite).

Reply:
I went to another approach, i.e. wrapping with `tokio::task::spawn_blocking` the top-level request to discover a single endpoint, instead of making these low-level methods async. I tried the latter approach at first, but I wasn't unable to make it work because `coap::Client` itself doesn't implement `Send`, so making every method that uses it async was very troublesome.